diffrax/_adjoint.py

```diff
@@ -852,3 +865,427 @@ def loop(
+@eqx.filter_custom_vjp
+def _loop_reversible(y__args__terms, *, self, throw, init_state, **kwargs):
+    del throw
+    y, args, terms = y__args__terms
+    init_state = eqx.tree_at(lambda s: s.y, init_state, y)
+    del y
+    return self._loop(
+        args=args,
+        terms=terms,
+        init_state=init_state,
+        inner_while_loop=ft.partial(_inner_loop, kind="lax"),
+        outer_while_loop=ft.partial(_outer_loop, kind="lax"),
+        **kwargs,
+    )
+
+
+@_loop_reversible.def_fwd
+def _loop_reversible_fwd(perturbed, y__args__terms, **kwargs):
+    del perturbed
+    final_state, aux_stats = _loop_reversible(y__args__terms, **kwargs)
+    ts = final_state.reversible_ts
+    ts_final_index = final_state.reversible_save_index
+    ys = final_state.save_state.ys
+    solver_state1 = final_state.solver_state
+    return (final_state, aux_stats), (ts, ts_final_index, ys, solver_state1)
+
+
+@_loop_reversible.def_bwd
+def _loop_reversible_bwd(
+    residuals,
+    grad_final_state__aux_stats,
+    perturbed,
+    y__args__terms,
+    *,
+    self,
+    saveat,
+    init_state,
+    solver,
+    event,
+    **kwargs,
+):
+    assert event is None
+
+    del perturbed, self, init_state, kwargs
+    ts, ts_final_index, ys, solver_state1 = residuals
+    original_solver_state, z1 = solver_state1
+    del residuals, solver_state1
+
+    grad_final_state, _ = grad_final_state__aux_stats
+    # If true we must be using SaveAt(t1=True).
+    t1_only = saveat.subs.t1
+    if t1_only:
+        y1 = (ω(ys)[-1]).ω
+        grad_y1 = (ω(grad_final_state.save_state.ys)[-1]).ω
+        grad_y1 = jtu.tree_map(_materialise_none, y1, grad_y1)
+        grad_y0_zeros = jtu.tree_map(jnp.zeros_like, grad_y1)
+
+    # Otherwise we must be using SaveAt(..., steps=True) due to the guard in
+    # ReversibleAdjoint. If y0 is not saved (t0=False) then we prepend grad_y0 (zeros).
+    else:
+        if saveat.subs.t0:
+            y1 = (ω(ys)[ts_final_index]).ω
+            grad_ys = grad_final_state.save_state.ys
+        else:
+            y1 = (ω(ys)[ts_final_index - 1]).ω
+            grad_ys = grad_final_state.save_state.ys
+            grad_y0 = jtu.tree_map(lambda x: jnp.zeros_like(x[0]), grad_ys)
+            grad_ys = jtu.tree_map(
+                lambda x, y: jnp.concatenate([x[None], y]), grad_y0, grad_ys
+            )
+
+        grad_ys = jtu.tree_map(_materialise_none, ys, grad_ys)
+        grad_y1 = (ω(grad_ys)[ts_final_index]).ω
+
+    del grad_final_state, grad_final_state__aux_stats
+
+    y, args, terms = y__args__terms
+    del y__args__terms
+
+    diff_args = eqx.filter(args, eqx.is_inexact_array)
+    diff_terms = eqx.filter(terms, eqx.is_inexact_array)
+    diff_z1 = eqx.filter(z1, eqx.is_inexact_array)
+    grad_args = jtu.tree_map(jnp.zeros_like, diff_args)
+    grad_terms = jtu.tree_map(jnp.zeros_like, diff_terms)
+    grad_z1 = jtu.tree_map(jnp.zeros_like, diff_z1)
+    del diff_args, diff_terms, diff_z1
+
+    def grad_step(state):
```

Comment:
Hmm, I'm trying to figure out how `solver_state` is handled here. I don't think it is correct?

`solver_state` is some completely arbitrary information that is propagated forward step-by-step, internal to the solver. In particular we don't have an API for reconstructing this backwards in time reversibly.